python/ray/tempfile_services.py

```diff
@@ -0,0 +1,261 @@
+import binascii
+import collections
+import datetime
+import errno
+import logging
+import os
+import shutil
+import tempfile
+
+import ray.utils
+
+logger = logging.getLogger(__name__)
+
+RAY_TEMPDIR = 'RAY_TEMPDIR'
+RAY_OBJECT_STORE_SOCKET_NAME = 'RAY_OBJECT_STORE_SOCKET_NAME'
+
+_incremental_dict = collections.defaultdict(lambda :0)
+
+
+def make_inc_temp(suffix="", prefix="", dir=None):
+    """Return a incremental temporary file name. The file is not created."""
+
+    if dir is None:
+        dir = tempfile.gettempdir()
+
+    index = _incremental_dict[suffix, prefix, dir]
+    for seq in range(tempfile.TMP_MAX):
+        index += 1
+        file = os.path.join(dir, prefix + str(index) + suffix)
+        if not os.path.exists(file):
+            _incremental_dict[suffix, prefix, dir] = index  # Save the index.
+            return file
+
+    raise FileExistsError(errno.EEXIST, "No usable temporary filename found")
+
+
+def try_to_create_directory(directory_path):
+    """Attempt to create a directory that is globally readable/writable.
+
+    Args:
+        directory_path: The path of the directory to create.
+    """
+    if not os.path.exists(directory_path):
+        try:
+            os.makedirs(directory_path)
+        except OSError as e:
+            if e.errno != os.errno.EEXIST:
+                raise e
+            logger.warning(
+                "Attempted to create '{}', but the directory already "
+                "exists.".format(directory_path))
+        # Change the log directory permissions so others can use it. This is
+        # important when multiple people are using the same machine.
+        os.chmod(directory_path, 0o0777)
+
+
+if RAY_TEMPDIR in os.environ:
+    temp_root = os.environ[RAY_TEMPDIR]
+    try_to_create_directory(temp_root)
+else:
+    # Change the default root by providing the directory named by the
+    # TMPDIR environment variable.
+    temp_root = tempfile.mkdtemp(prefix='ray_{pid}_'.format(pid=os.getpid()))
```

Comment:
On my macbook, `tempfile.mkdtemp()` returns `'/var/folders/15/54jf68993rd7753c5fms424r0000gn/T/tmpwtgcu22l'`. This seems substantially less usable than `/tmp/raylogs` because no one will be able to find the directory.

Reply:
This is the [algorithm Python uses for temp dir](https://docs.python.org/3/library/tempfile.html#tempfile.gettempdir).

Reply:
It feels useful to me to have a consistent directory so people know where to look even if they lost the stdout from `ray.init`. It'll have to be a little different on windows, but that doesn't seem too bad.

Reply:
So `mkdtemp` and other methods do get `dir` as an argument where you can set this directory. So then I suggest this is a command line argument which is by default `/tmp` but users can change it. But still, that secure `mkdtemp` and other functions are used, and that it is configurable.